fearless/Modules/Staking/StakingPayoutConfirmation/StakingPayoutConfirmationPresenter.swift

```diff
@@ -1,13 +1,137 @@
 import Foundation
+import BigInt
+import CommonWallet
 
 final class StakingPayoutConfirmationPresenter {
     weak var view: StakingPayoutConfirmationViewProtocol?
     var wireframe: StakingPayoutConfirmationWireframeProtocol!
     var interactor: StakingPayoutConfirmationInteractorInputProtocol!
+
+    private var fee: Decimal?
+    private var priceData: PriceData?
+    private var balance: Decimal?
+
+    let balanceViewModelFactory: BalanceViewModelFactoryProtocol
+    let logger: LoggerProtocol?
+    let asset: WalletAsset
+
+    init(
+        balanceViewModelFactory: BalanceViewModelFactoryProtocol,
+        asset: WalletAsset,
+        logger: LoggerProtocol? = nil
+    ) {
+        self.balanceViewModelFactory = balanceViewModelFactory
+        self.asset = asset
+        self.logger = logger
+    }
 }
 
 extension StakingPayoutConfirmationPresenter: StakingPayoutConfirmationPresenterProtocol {
-    func setup() {}
+    func setup() {
+        provideFee()
+
+        interactor.setup()
+        interactor.estimateFee()
+    }
+
+    func proceed() {
+        guard let fee = fee else {
+            if let view = view {
+                wireframe.presentFeeNotReceived(
+                    from: view,
+                    locale: view.localizationManager?.selectedLocale
+                )
+            }
+
+            return
+        }
+
+        interactor.submitPayout(for: balance ?? 0.0, lastFee: fee)
+    }
+
+    // MARK: - Private functions
+
+    private func provideFee() {
+        if let fee = fee {
+            let viewModel = balanceViewModelFactory.balanceFromPrice(fee, priceData: priceData)
+            view?.didReceive(feeViewModel: viewModel)
+        } else {
+            view?.didReceive(feeViewModel: nil)
+        }
+    }
+
+    private func handle(error: Error) {
+        let locale = view?.localizationManager?.selectedLocale
+
+        if let confirmError = error as? StakingConfirmError {
```

Comment:
I think it should be different errors for payouts. At least we don't have "amount too high" and "missing controller" cases.

Reply:
Agreed. Actually, I thought about it, but decided to live as is for a moment. Since we have many confirmation screens with many check, do you have any ideas of how to make it more universal?

Reply:
BTW, we don't have "amount too high" in StakingConfirmError as well, should it be there?